### ISSUE TITLE
Allowed at-type version differences in same pipeline

### ISSUE THREAD

**Issue by marcverhagen** [clamsproject/mmif-python#163]
Now a component looking for NamedEntities that is using MMIF 0.3.1 will look for http:/mmif.clams.ai/versions/0.3.1 in views and miss any entity that is of MMIF version 0.3.0.

**Comment by keighrim** [comment]
This directly relates to discussion in https://github.com/clamsproject/mmif/issues/14 . But the original issue is somewhat outdated by now, and I think a good place to start to update the discussion is to define what *breaking* changes are, using our experience past year or so.

**Comment by marcverhagen** [comment]
Yes, it does relate, and I think it is a good idea to cast the discussion in terms of what *breaking* really means, at least for now. I see several levels of breakage associated with the major-minor-patch changes:

1. The JSON schema are different. This breaks the syntax and could, and probably even should, throw an error. We agreed that this could be the thing that gets us to bump up the major version once we finally hit version 1.0.0. A complication here is that some changes will invalidate things that used to be okay and others won't. For example, making a property optional won't break anything, making it required would, at least for all the cases where we do not have that property in the MMIF file.

2. We add, remove or change an annotation type (where changing means adding, removing or changing a property or metadata property). We had been thinking of these as changes to the minor version. Here we could glue together a chain with and EAST version that uses MMIF 1.2.0 and Tesseract that uses MMIF 1.3.0. This probably won't break the chain unless an app is set up to throw an error when required input isn't there (and I think the app should in that case set up to be really specific about what kind of input it misses and put that in the error message). Glueing together these components may break the intended result if there are enough changes in MMIF pertinent to the annotation types that EAST and Tesseract use. For example, Tesseract would return an empty result because it did not find and bounding boxes of the right version.

3. No breakage at all. Any pipeline runs exactly as it did before the MMIF change. When MMIF changes at this patch-level nothing else needs to change.

Loose thoughts:

- The hard thing is to decide what to do with the second case.
- App metadata never refer to the version of a required annotation type, which is as it should be, the question is whether the MMIF version used by the app not only determines what the app creates but also what it needs as input.
- I don't like at all the idea of all applications needing to be changed following a minor change in the specifications, even if those changes to not impact the app at all. This would be unmanageable.
- Maybe `ClamsApp.annotate()` in `clams-python` could throw an error when the app's major MMIF version conflicts with the major MMIF version in the input MMIF file.
- The `clams-python` package is set up to require a specific X.Y.Z version of the `mmif-python` package. I don't know what exactly the nature of this dependency is. Would it be feasible to use a range? At least for the patch-level and maybe also for the minor version.

Maybe this should be in https://github.com/clamsproject/mmif/issues/14.

**Comment by keighrim** [comment]
For implementing this, I'd also like to propose we re-visit https://github.com/clamsproject/mmif-python/issues/23 (and possibly https://github.com/clamsproject/mmif-python/issues/64). See [comment ](https://github.com/clamsproject/mmif-python/issues/23#issuecomment-847224499) for more details.

**Comment by keighrim** [comment]
Here are places where the information / documentation on versioning is written outside of this repository. They also need to be updated when this issue is addressed. 

* https://mmif.clams.ai/versioning/
* https://sdk.clams.ai/introduction.html#note-on-versions

**Comment by keighrim** [comment]
`versioning` documentation is updated via https://github.com/clamsproject/mmif/pull/177, and will be released soon. Closing this issue.

**Comment by keighrim** [comment]
reopening this issue as we are now discussing allowing any versions in the same pipeline.

**Comment by keighrim** [comment]
The primary argument for allowing "any" version of `@type` in a pipeline (to my understanding) is that even when we have major/minor-different vocab versions, an annotation type may have stayed the same for all the updates in the vocab. For example, `https://blah.blah/a.b.c/TimeFrame` and `https://blah.blah/x.y.z/TimeFrame` can be precisely the same because all the updates between `a.b.c.` and `x.y.z` were irrelevant to the `TimeFrame` type. In other words, since we don't use separate version numbers to keep track of changes of individual `@type`s (we just have one version number of the whole type hierarchy), we can't tell if different MMIF/vocab version numbers mean break-or-not for particular annotation types.

**Comment by marcverhagen** [comment]
Adding incompatibility warnings definitely sounds like a good idea. But I don't think I get the workflow example because it includes a CLAMS App passing a warning and I don't think warnings should be passed to the next view. 

Question on this prose:

> we can generate, for each of `@type`, a list of MMIF versions that contains any changes to the `@type` and at the build time and encode that list inside the AnnotationType enum in the release package.

Are you saying here we determine equivalence classes of MMIF versions for each `@type`? 

One more thought on the requirements. I wonder how involved it would be to include a pip-like requirement specifier, where instead of 

```python
metadata.add_input(DocumentTypes.BoundingBox)
```

we would use

```python
metadata.add_input(DocumentTypes.BoundingBox, version=('>=0.34', '<=0.5.0')
```

Changing the `add_input` method is easy enough but that would not be the only change. And we would have to think about defaults and what happens when new versions pop up.  Maybe some mixture of this with calculating what identical versions are.

**Comment by keighrim** [comment]
> But I don't think I get the workflow example because it includes a CLAMS App passing a warning and I don't think warnings should be passed to the next view.

My example of "data flow" was not an example of a clams workflow, but rather an in-memory information flow within a single clams app. What I tried to say was that the warnings are issued by the MMIF parse code, passed through python APIs, and finally serialized by the common `clams-python` API `annotate()`, so that individual developer doesn't have to bother with the warnings in their actual implementation of the data processing in `_anntate()` method. 

---
> Are you saying here we determine equivalence classes of MMIF versions for each `@type`? 

Yes. Here's a specific example; 
Suppose, among many others, we have two `at_type`s, `uri/x.y.z/TypeA` (or `TypeA` in short) and `uri/x.y.z/TypeB` (or `TypeB`). Along the way we update our MMIF spec, we make some changes to those two in these versions 

| MMIF version | notable changes |
|---|---|
|1.0.0|init release only with `TypeA`|
| ... | | 
|1.2.0|added `TypeB` |
| ... | | 
|1.3.0 | added a required field to `TypeB`|
| ... | | 
|1.6.0 | added a required field to `TypeA` | 
| ... | | 
|2.0.0 | added a required field to all `at_type`s | 
| ... | | 

1. Now, at the very beginning, one developer writes a clams app that generates `TypeA`. Initially, the app is based on MMIF 1.0.0, so let's call it `app_a_generator`, and versioned as `v1`. This app will create a view that `contains` `uri/1.0.0/TypeA`, where the `/1.0.0/` part of the `at_type` string is coming from the `mmif.__specver__` of the SDK that the app is based on.
1. Then after MMIF 1.1.0 (and related `mmif-python` and `clams-python`) is released. The developer updates the app to use "the latest". This is `app_a_generator:v2` (colon as a version delimiter) and creates `uri/1.1.0/TypeA`-typed annotation objects.
1. `uri/1.0.0/TypeA` and `uri/1.1.0/TypeA` are semantically (and very likely syntactically) the same types, because there have been no changes regarding `TypeA` in the MMIF spec between the two versions. 
1. Again later, MMIF 1.2.0 (and SDKs) is released and `app_a_generator` is also updated to `v3`, targeting MMIF 1.2.0. 
1. Around the same time, another clams app based on MMIF 1.2.0 is developed. It takes `TypeA` as input and generates `TypeB` as output. Let's call it `app_a_to_b:v1`. Now, this one generates views that `contains` `uri/1.2.0/TypeB` annotations. 
    * For the input specification, the developer will have something like this in the `_appmetadata()` method. 
        ``` python 
        metadata.add_input(AnnotationTypes.TypeA)
        ```
    * When we run a clams pipeline that chains `app_a_generator:v3` `->` `app_a_to_b:v1`, the input to `app_a_to_b` is `uri/1.2.0/TypeA` and the app is based on MMIF 1.2.0 and thus generates `uri/1.2.0/TypeN`. There shouldn't be a problem with it. 
    * Similarly, when we run a clams pipeline that chains `app_a_generator:(v1|v2)` `->` `app_a_to_b:v1`, the input to `app_a_to_b` is `uri/1.(0|1).0/TypeA` but the app generates `uri/1.2.0/TypeB`. But because `uri/1.0.0/TypeA`, `uri/1.1.0/TypeA`, and `uri/1.2.0/TypeA` are all the same object with slightly different type names, this shouldn't be a problem. 
1. After a while, MMIF 1.6.0 is released and it makes a *breaking* change to `TypeA`. And soon, targeting the new MMIF spec, `app_a_generator` is updated to `v4` and `app_a_to_b` to `v2`. 
    * Note that because of the change made to `TypeB` in MMIF 1.3.0, `app_a_to_b:v2` might need a serious implementational change to correctly work.
    * At this point, if one chains `app_a_generator:v1` `->` `app_a_to_b:v2`, the app must know there could be problems, and thus the resulting MMIF file will contain a view with warnings, based on the internal "magic" that knows that `uri/1.0.0/TypeA` (from the input file) is a different creature from `uri/1.6.0/TypeA` (from the object defined in the `mmif-python` in the site-package). 
    * More concretely, the "magic" is just a set of integer tuples inside the `AnnotationType.TypeA` class that contains all the points (MMIF version number) when a change's made to `TypeA`. In this case, there's only one element `(1,6,0)`. This set can be automatically generated at the build time by "diff"ing the vocab files from all previous releases. 
    * If we don't want the magic and want to make everything transparent and explicit, I believe the only way is to have separate versioning schemes for all individual `at_type`s. 

---
> what happens when new versions pop up

This approach can automate the determination of lower boundaries of the equivalence classes of `at_type`s. We still have a problem with the upper bounds. In the above example, imagine that one chains `app_a_generator:v4` (based on MMIF 1.6.0) `->` `app_a_to_b:v1` (based on MMIF 1.2.0). From the perspective of `mmif-python:1.6.0`, it knows that `uri/1.6.0/TypeA` and `uri/1.2.0/TypeA` are the "different" types, but `mmif-python:1.2.0` doesn't know that as it didn't have access to the future at the build time. Hence this can be a problem. 1.2.0 SDK can check the online vocab website to determine the equivalence/identity/compatibility, but it requires the network connection at the runtime, and hence not so desirable. 

One way to solve the upper bound issue it to use the "lower bound" number in the output MMIF. For instance, `app_a_generator:v2` (based on MMIF 1.1.0) can create `uri/1.0.0/TypeA`-typed annotation objects (instead of using `mmif.__specver__`, which is `/1.1.0/`) because the underlying SDK knows `uri/1.1.0/TypeA` and `uri/1.0.0/TypeA` are identical. In the same way, `app_a_generator:v3` creates `uri/1.0.0/TypeA`, `app_a_generator:v4` creates `uri/1.6.0/TypeA`. By using the exact same numbers for semantically identical types, SDKs can use simple string matches to determine two types with the same suffix are identical or not. Namely, all equivalence classes of at_types are now singletons and we don't have to worry about the range of versions. Note that we still need the same "magic" to determine the lower-bound numbers to embed in the output MMIF.

---
> One more thought on the requirements. I wonder how involved it would be to include a pip-like requirement specifier, 

The "singleton-equivalence-class (SEC)" method described above would simplify the problem down to a simple string match. I'm not so sure if adding a complex range-based version specification in the input/requirement app_metadata can replace the SEC method or provide additional solutions that the SEC can't handle.

**Comment by keighrim** [comment]
In an in-person meeting today, Marc and I talked about moving to a versioning model where all individual annotation types have their own version schemes, decoupled from the MMIF version. The necessity came up when we were discussing a possible way to allow an app to use "very" old generations of an `@type` that's no longer supported (or cared for) in the MMIF that the SDK is based on. 
We concluded until we have a better idea of how to move to the new model, and assess how that will affect existing apps and users (primarily GBH), we don't touch the current compatibility check mechanism, which simply relies on string matches of major, minor numbers.

**Comment by keighrim** [comment]
Just for reference for the future discussion, current apps listed in [clams-apps](https://github.com/clamsproject/clams-apps/blob/7a8b6d4b4d6edf01ad5bcbe41288c41bebb8f3cf/README.md) are all support at least MMIF 0.4.0. So we don't have to worry about back-supporting MMIF 0.3.0 or older moving forward.

**Comment by keighrim** [comment]
So this is turning into three different problems now. 

1. Defining *eequivalence classes* of annotation types.  This can be done either under an umbrella MMIF version number or individuated versioning schemes. More on this should be discussed in https://github.com/clamsproject/mmif/issues/14
1. Whether to treat different versions of individual annotation types as monotonic *improvements* (e.g. `uri/v2/TypeA` is a better version of `uri/v1/TypeA`, thus once `v2/TypeA` is out, `v1/TypeA` should be deprecated and usage of it should be discouraged) or treat them as just *differently defined* objects (e.g. `v2/TypeA` is just differently defined `TypeA`, and developers have the freedom to choose between `v2/TypeA` and `v1/TypeA`). 
1. Implement how to enforce the type-checking in a CLAMS pipeline. If we encourage using the latest versions of annotation types all the time, I believe my proposal above is the best solution to this problem so far. On the other hand, if we allow developers to freely choose which version(s) of input types their app accepts, I think the best way to conduct the type-checking is not by using *numbering* but instead using some kind of *dynamic typing* based on the properties of objects. 

One more thing: the reason we need to think about the second problem. In the above example with `app_a_generator` and `app_a_to_b`, when MMIF 1.6.0 is released and `app_a_generator:v4` and `app_a_to_b:v2` are also released based on 1.6.0 spec, the *current generation* at_types are `uri/1.6.0/TypeA` and `uri/1.3.0/TypeB`. Thus, `app_a_to_b:v2` should issue warnings, or even worse might completely failed to process, when given `uri/1.0.0/TypeA`-likes. But what if a user already already spent thousands of hours processing thousand of media files using `app_a_generator:v3` (or older) and have thousands of MMIF files that contain `uri/1.0.0/TypeA`-likes? The user probably don't want to spend another thousands of hours running `app_a_generator:v4` on the source files. Thus, a developer might want to develop `app_a_to_b:v3` that takes `uri/1.0.0/TypeA` as input but still wants to use `mmif-python@1.6.0`.

**Comment by keighrim** [comment]
From the in-person discussion today, I think we agreed that supporting "free-range" version specification in app metadata is a bad idea. If an app developer *ever* needs to write an app based on the latest SDK but targeting an old at_type, writing a separate "converter" (old at_type `properties` structure -> new at_type structure) should be considered. 

---
Given that, my conclusion; 

1. SDK won't support "ranged" version specifications in input requirement in `appmetadata`, neither do individual apps.
2. SDK won't support "dynamic typing" of at_type structures in any foreseeable future. For now, type-checking in a pipeline engine should only rely on the at_type name and the at_type version.
3. Choice between "umbrella" versioning and "individuated" versioning of at_types has not been made, but we all agree that at_type versions should be more-or-less independent of the SDK version. 

---
On the 3rd point, I believe "umbrella" versioning should be easier implementation-wise and also semantically it's better because it presents the originating information of the at_type (which CLAMS vocab tree a type came from) more transparently, and thus we can make explicit links to a vocab tree. So my vote is for the umbrella scheme. I want to hear what others think.
